### ISSUE TITLE
SelfService user is unable to revoke a verified second factor

### ISSUE THREAD

**Issue by phavekes** [OpenConext/Stepup-SelfService#416]
[This issue is imported from pivotal - Originaly created at May 7, 2019 by Michiel Kodde](https://www.pivotaltracker.com/story/show/165850321)

# Description
A self service user tries to remove a verified token that has expired. And is presented with the following exception: `Error: Call to a member function isSraa() on null`. The user is not a privileged (SRA(A)) user. 

SelfService verified that the user is the owner of the token and sends the Revoke command to Middleware. Middleware is not able to assert that the user may apply the command. 

# Diagnosis
Right befor the command is executed, Middleware performs a final assertion if the user is allowed to perform the action. This is done by calling the [`assertCommandMayBeProcessedOnBehalfOf`](https://github.com/OpenConext/Stepup-Middleware/blob/209ceecec6c690636b9e451ab8032c424dcc875d/src/Surfnet/StepupMiddleware/ApiBundle/Controller/CommandController.php#L152) method. 

This method first checks if the institution of the user is on the whitelist. If this is the case, the command may be applied. If this is not the case, a fallback check is performed. If the user is SRAA, the command is still valid.

In the case of this bug report, the institution of the user is not whitelisted, and the user is not privileged.

# Environment
Pilot environment: Release 16 is installed
MW version: 2.9.3 (31edf17a2d3dcf7f5df54f385926fcb5833c76c1)
SS version: 2.10.6 (945e18d66d3763752faae3dbec06874630224509)

# Proposal
1. The verification whether or not the user [is a SRAA](https://github.com/OpenConext/Stepup-Middleware/blob/209ceecec6c690636b9e451ab8032c424dcc875d/src/Surfnet/StepupMiddleware/ApiBundle/Controller/CommandController.php#L167) should be made safe. Not all users are listed in the RA listing (only (S)RA(A) users are). So checking if `registrationAuthorityCredentials` is `null` is good practice here. Preventing the error to begin with.
2. The actual problem lies in the fact that the institution is not present in the whitelist projection. The institution was whitelisted in the config (see logs in the Slack thread). But was not projected successfully, or was later dropped during another whitelist action. This needs to be investigated and possibly fixed.
[Update by @pmeulen: The institution was removed from the whitelist, so it is correct that it does not appear on the `whitelist_entry` projection]

# Discussion
The Middleware assertion ([`assertCommandMayBeProcessedOnBehalfOf`](https://github.com/OpenConext/Stepup-Middleware/blob/209ceecec6c690636b9e451ab8032c424dcc875d/src/Surfnet/StepupMiddleware/ApiBundle/Controller/CommandController.php#L152)) looks to be created with RA in mind. It also works for SelfService as long as the institution for the user is whitelisted. Is this safe enough? Or do we want separate assertions for SS and RA?

Let\'s discuss in the comment section below!

**Comment by phavekes** [comment]
Some concerns mentioned on Slack by  @pmeulen:

1. The user should not be able to log in to self service as the institution is no longer on the whitelist
2. The create identiy and token registration events where applied **after** the institution was removed from the whitelist. MW should have rejected those events.

@bstrooband also mentioned a possible issue:
1. Updates to the whitelist are not reflected in the institution listing projection. So institutions that no longer are whitelisted can still be listed in the application, for example in the institution switchers we introduced in the RA env. (Michiel Kodde - May 7, 2019)

**Comment by phavekes** [comment]
* I saved a database dump, so this can be consulted later if required.

* @bstrooband regarding updating the institution listing projection when an institution is removed from the whitelist: assuming the `institution_listing` projection is only used for filling the institution selection in the RA interface, I think that it is correct that the removed institution is still in that projection. As an SRAA I want still be able to consult the token overview en especially the audit log of an institution that was removed from the whitelist.

* That we are have a discussion of whether something should be present in a projection or not is an indication that it is currently not sufficiently clear what the semantics of the projections are. I think it is important that we make the contract between the projections and the code using the projections clear.

* After verifying the event log I am sure that the identity creation and token registration events where applied while the institution was still on the whitelist. The MW was thus correct in accepting those commands.

* After adding the institution to the whitelist again, the token could be revoked and a new one could be registered. (Pieter van der Meulen - May 7, 2019)
